engine/consensus/matching/engine.go

```diff
@@ -908,15 +930,97 @@ func (e *Engine) requestPending() error {
 		blockID := header.ID()
 		if _, ok := knownResultForBlock[blockID]; !ok {
 			missingBlocksOrderedByHeight = append(missingBlocksOrderedByHeight, blockID)
+			continue
 		}
+
 	}
 
 	// request missing execution results, if sealed height is low enough
 	log.Info().
 		Int("finalized_blocks_without_result", len(missingBlocksOrderedByHeight)).
 		Msg("requesting receipts")
 	for _, blockID := range missingBlocksOrderedByHeight {
-		e.requester.EntityByID(blockID, filter.Any)
+		e.receiptRequester.EntityByID(blockID, filter.Any)
+	}
+
+	return nil
+}
+
+// requestPendingApprovals requests approvals for chunks that haven't collected
+// enough approvals. When the number of unsealed finalized blocks exceeds the
+// threshold, we go through the entire mempool of incorporated-results, which
+// haven't yet been sealed, and check which chunks need more approvals.
+func (e *Engine) requestPendingApprovals() error {
+	// last sealed block
+	sealed, err := e.state.Sealed().Head()
+	if err != nil {
+		return fmt.Errorf("could not get sealed height: %w", err)
+	}
+
+	// last finalized block
+	final, err := e.state.Final().Head()
+	if err != nil {
+		return fmt.Errorf("could not get finalized height: %w", err)
+	}
+
+	// only request if number of unsealed finalized blocks exceeds the threshold
+	log := e.log.With().
+		Uint64("finalized_height", final.Height).
+		Uint64("sealed_height", sealed.Height).
+		Uint("sealing_threshold", e.sealingThreshold).
+		Logger()
+	if uint(final.Height-sealed.Height) < e.sealingThreshold {
+		log.Debug().Msg("skip requesting approvals as number of unsealed finalized blocks is below threshold")
+		return nil
+	}
+
+	log.Info().Msg("requesting approvals")
+	for _, r := range e.incorporatedResults.All() {
+		resultID := r.Result.ID()
+
+		// Compute the chunk assigment. Chunk approvals will only be requested
+		// from verifiers that were assigned to the chunk. Note that the
+		// assigner keeps a cache of computed assignments, so this is not
+		// necessarily an expensive operation.
+		assignment, err := e.assigner.Assign(r.Result, r.IncorporatedBlockID)
+		if state.IsNoValidChildBlockError(err) {
+			continue
+		}
+		if err != nil {
+			// at this point, we know the block and a valid child block exists.
+			// Not being able to compute the assignment constitutes a fatal
+			// implementation bug:
+			return fmt.Errorf("could not determine chunk assignment: %w", err)
+		}
+
+		// send approval requests for chunks that haven't collected enough
+		// approvals
+		for _, c := range r.Result.Chunks {
+
+			// skip if we already have one valid approval for this chunk.
+			// TODO: this is the happy path. In the full protocol we will
+			// require more than one approval.
+			sigs, ok := r.GetChunkSignatures(c.Index)
```

Comment:
Since the protocol doesn't have a way to say "this result is invalid, and there is no approval for it", if a result is indeed invalid, then the consensus nodes would have to keep requesting approvals for it until the block is sealed with a different result. Right?

Reply:
Yes that is a good point. 

We will keep requesting approvals for "bad" results until the corresponding block is sealed, because right now, the only way for a verification node to say they do not approve a Result is to not send an approval.

Would you recommend implementing some sort of counter whereby we never re-request the same approval more than x times? That could be a quick hack.

Otherwise is there a plan to add a concept of "Refusal" to the verification nodes, whereby they could send an explicit refusal ( even if just a new bool field in ResultApproval) ? After collecting enough refusals ( the other portion of nodes than is required for sealing ) we could remove the result from mempool, and possibly challenge the ExecutionNode?

Reply:
ok, following our discussion, I implemented the following rate-limiting system:

https://github.com/onflow/flow-go/blob/4b91a746a6fa80dda2ad3d0d18d3c5f8bb657638/engine/consensus/matching/engine.go#L1038-L1050

The `RequestTracker` is basically a wrapper around a map that tracks the number of requests per result/chunk-index, and it gets cleaned up in the `clearPools` function:

https://github.com/onflow/flow-go/blob/4b91a746a6fa80dda2ad3d0d18d3c5f8bb657638/engine/consensus/matching/engine.go#L854-L860